compiler/src/vm/tracer.rs

```diff
@@ -140,67 +175,226 @@ impl Tracer {
+                            args: args.clone(),
+                        }),
+                        inner: vec![],
+                    });
                 }
-                TraceEntry::CallStarted { id, closure, args } => Action::Start(format!(
-                    "{id} {} {}",
-                    closure.format(heap),
-                    args.iter().map(|arg| arg.format(heap)).join(" ")
-                )),
-                TraceEntry::CallEnded { return_value } => {
-                    Action::End(format!(" = {}", return_value.format(heap)))
+                EventData::CallEnded { return_value } => {
+                    let span = stack.pop().unwrap();
+                    let (id, closure, args) = match span.data.unwrap() {
+                        StackEntry::Call { id, closure, args } => (id, closure, args),
+                        _ => unreachable!(),
+                    };
+                    stack.last_mut().unwrap().inner.push(Trace {
+                        start: span.start,
+                        end: event.when,
+                        data: TraceData::Call {
+                            id,
+                            closure,
+                            args,
+                            inner: span.inner,
+                            result: TraceResult::Returned(*return_value),
+                        },
+                    });
                 }
-                TraceEntry::NeedsStarted {
+                EventData::NeedsStarted {
                     id,
                     condition,
                     reason,
-                } => Action::Start(format!(
-                    "{id} needs {} {}",
-                    condition.format(heap),
-                    reason.format(heap)
-                )),
-                TraceEntry::NeedsEnded => Action::End(" = Nothing".to_string()),
-                TraceEntry::ModuleStarted { module } => Action::Start(format!("module {module}")),
-                TraceEntry::ModuleEnded { export_map } => Action::End(export_map.format(heap)),
-            })
-            .collect_vec();
-
-        let mut lines = vec![];
-        let mut stack = vec![];
-        let mut indentation = 0;
-
-        for action in actions {
-            let indent = "  ".repeat(indentation);
-            match action {
-                Action::Start(line) => {
-                    stack.push(lines.len());
-                    lines.push(format!("{indent}{line}"));
-                    indentation += 1;
+                } => {
+                    stack.push(Span {
+                        start: event.when,
+                        data: Some(StackEntry::Needs {
+                            id: id.clone(),
+                            condition: *condition,
+                            reason: *reason,
+                        }),
+                        inner: vec![],
+                    });
+                }
+                EventData::NeedsEnded { nothing } => {
```

Comment:
Instead of storing `Nothing` for every `::NeedsEnded` just for this purpose, we could store it just once, e.g., in the tracer. However, this might be premature optimization. Also, since `::NeedsEnded` is still smaller than, e.g., `::NeedsStarted`, the enum's size in memory won't change

Reply:
That's true. In the next PR that rearchitects the tracing, I fix this problem, so I won't change this here.